pkg/services/ngalert/migration/persist.go

```diff
@@ -0,0 +1,324 @@
+	generalAlertingFolder *folder.Folder
+}
+
+// newSync creates a new migrationService for the given orgID.
+func (ms *migrationService) newSync(orgID int64) *sync {
+	return &sync{
+		orgID: orgID,
+		log:   ms.log.New("orgID", orgID),
+
+		migrationStore: ms.migrationStore,
+		getDecryptedValue: func(ctx context.Context, sjd map[string][]byte, key, fallback string) string {
+			return ms.encryptionService.GetDecryptedValue(ctx, sjd, key, fallback)
+		},
+		channelCache: ms.newChannelCache(orgID),
+
+		permissionsMap:        make(map[int64]map[permissionHash]*folder.Folder),
+		folderCache:           make(map[int64]*folder.Folder),
+		folderPermissionCache: make(map[string][]accesscontrol.ResourcePermission),
+	}
+}
+
+// syncAndSaveState persists the given dashboardUpgrades and contactPairs to the database.
+func (sync *sync) syncAndSaveState(
+	ctx context.Context,
+	dashboardUpgrades []*migmodels.DashboardUpgrade,
+	contactPairs []*migmodels.ContactPair,
+) error {
+	delta, err := createDelta(dashboardUpgrades, contactPairs)
+	if err != nil {
+		return fmt.Errorf("state delta: %w", err)
+	}
+
+	state, err := sync.syncDelta(ctx, delta)
+	if err != nil {
+		return fmt.Errorf("sync state: %w", err)
+	}
+
+	err = sync.migrationStore.SetOrgMigrationState(ctx, sync.orgID, state)
+	if err != nil {
+		return fmt.Errorf("save state: %w", err)
+	}
+
+	return nil
+}
+
+// StateDelta contains the changes to be made to the database based on the difference between
+// existing migration state and new migration state.
+type StateDelta struct {
+	DashboardsToAdd []*migmodels.DashboardUpgrade
+	ChannelsToAdd   []*migmodels.ContactPair
+}
+
+// createDelta creates a StateDelta from the new dashboards upgrades and contact pairs.
+func createDelta(
+	dashboardUpgrades []*migmodels.DashboardUpgrade,
+	contactPairs []*migmodels.ContactPair,
+) (StateDelta, error) {
+	return StateDelta{
+		DashboardsToAdd: dashboardUpgrades,
+		ChannelsToAdd:   contactPairs,
+	}, nil
+}
+
+// syncDelta persists the given delta to the state and database.
+func (sync *sync) syncDelta(ctx context.Context, delta StateDelta) (*migrationStore.OrgMigrationState, error) {
+	state := &migrationStore.OrgMigrationState{
+		OrgID:          sync.orgID,
+		CreatedFolders: make([]string, 0),
+	}
+
+	amConfig, err := sync.handleAlertmanager(ctx, delta)
+	if err != nil {
+		return nil, err
+	}
+
+	err = sync.handleAddRules(ctx, state, delta, amConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return state, nil
+}
+
+// handleAlertmanager persists the given channel delta to the state and database.
+func (sync *sync) handleAlertmanager(ctx context.Context, delta StateDelta) (*migmodels.Alertmanager, error) {
+	amConfig := migmodels.NewAlertmanager()
+
+	if len(delta.ChannelsToAdd) == 0 {
+		return amConfig, nil
+	}
+
+	for _, pair := range delta.ChannelsToAdd {
+		if pair.ContactPoint != nil && pair.Route != nil {
```

Comment:
My first question was "what if only one is specified, and why we skip it?" Digging through the changes I found that it seems to be only possible when field Error is not empty. I wonder if we can make it more clear here by 
1. checking if Error is empty (or create a method IsValid)
2. handle ContactPoint and Route separately, ie `if pair.ContactPoint != nil { add receiver }; if pair.Route != nil { add route }.
I think this will be more straightforward.

Reply:
Yes, this is a bit awkward because we're collecting errors but not doing anything with them yet. This will change in the PR that starts using them in a similar way to what you're saying.